### PR TITLE
WebSocket methods acceptWithActor and tryAcceptWithActor should take …

### PR DESCRIPTION
**Pull request by ghost**
…an implicit ActorSystem, rather than an implicit Application. This allows the ActorSystem to be supplied via dependency injection, or for the user to supply their own. Fixes issue #4614 for compile-time dependency injection users.

### CODE REVIEW

**Comment by emergentcypher** [comment]
I've attempted to fix `WebSocket` for compile-time dependency injection users and allow
for an `ActorSystem` to be supplied by the caller, and prepare to move away from global state. For backwards compatibility, 
I think it would be best if `WebSocket.tryAcceptWithActor` could accept either 
an `Application`, or an `ActorSystem`. Implicit conversions are discouraged, but since the application parameter is an
implicit it can't be overloaded, so a simple way to do this would be via the magnet 
pattern. What I suggest is for the methods to accept a new `ApplicationOrActorSystem` 
object defined inside `WebSocket`, which can be implicitly built from either an `Application` or an 
`ActorSystem`.

This way people who still depend on `play.api.Play.current` can use the global state, 
while those of us moving on to either Guice or compile-time dependency injection can properly
supply an `ActorSystem`. Ideally this would eventually be updated to strictly require
an implicit `ActorSystem`.

**Comment by gmethvin** [comment]
I think it'd be best to keep binary compatibility with (and maybe deprecate) the existing method and introduce a new api for this. Maybe something like

```scala
// val webSocketApi = WebSocketApi(actorSystem)
webSocketApi.acceptWithActor { ... }
```

**Comment by emergentcypher** [comment]
Hmm, I think I agree, I prefer the parameter being explicit actually, and this way the WebSocketApi could be instantiated in the loader. So my controller could look like:

    class Application(webSocketApi: WebSocketApi) extends Controller {
      def socket = webSocketApi.acceptWithActor[JsValue, JsValue] { request =>
        EchoActor.props
      }
    }

And we don't need touch the original WebSocket code at all. So I updated my pull request. Thoughts? Or do you (or anyone) think this requires significantly more discussion and I should pull the pull request.

**Comment by gmethvin** [comment]
@jroper Any thoughts? (I think you wrote the websocket code originally).

**Comment by jroper** [comment]
I wonder if a better name for `WebSocketApi` is `WebSocketBuilder`, since that's what it is, a builder for web socket actions.  This is similar to `ActionBuilder`.

And then yes, it should probably be injected, which means bindings need to be provided for it, and the documentation should be updated to use the injected websocket builder.

**Comment by stsmedia** [comment]
I have been running into the same issue (I have my own actor system to allow for dynamic cluster seed node setup in EC2) but hit the wall with Play Websockets which has no support for external akka systems. As a workaround I have created my own WebSocketActor implementation. Obviously this is far from ideal.

Any plans when and if this will be released into 2.4.x?

**Comment by gmethvin** [comment]
We should release this into 2.4.x since this is, in my opinion, fixing a bug if 2.4. One of the goals of 2.4 was to replace features that relied on global state with injectable components, and somehow we missed this API.

**Comment by jroper** [comment]
Agreed.  @emergentcypher, what do you think of the name, WebSocketBuilder?

**Comment by gmethvin** [comment]
`WebSocketBuilder` seems like a good name.

We might want to move the other methods (`tryAccept`, `using`, `adapter`) there as well, even though there are no dependencies. You should also create a class (something like `DefaultWebSocketBuilder`) with an @Inject annotation on the constructor and a module for runtime DI.

To avoid code duplication you can add a builder to the `WebSocket` object
```scala
private def builder(implicit app: Application): WebSocketBuilder =
  new DefaultWebSocketBuilder(Akka.system)
```
and simply delegate to the builder methods.

**Comment by emergentcypher** [comment]
That sounds good to me--`WebSocketBuilder` makes sense. OK, I'll work on those changes and update the pull request.

**Comment by dkharrat** [comment]
Just came across this pull request after also encountering exceptions using WebSockets in play 2.4 using compile-time dependency injection.

Any update on this issue?

**Comment by julienrf** [comment]
Here is a summary of what I think is needed for this issue:

- [ ] Turn the `WebSocket` object into a `WebSocketBuilder` class that takes an `ActorSystem` as constructor parameter ;
- [ ] Add a `WebSocketBuilderModule` class for JSR-330 based DI and a `WebSocketBuilderComponents` trait for compile-time DI (that is, the trait should have an abstract `def actorSystem: ActorSystem` and a `lazy val webSocketBuilder = new WebSocketBuilder(actorSystem)` as members) ;
- [ ] Deprecate the existing `WebSocket` object.

**Comment by gmethvin** [comment]
@julienrf Good summary, and I like your use of checkboxes :+1: 

@emergentcypher any progress on this? It seems like we are on the right track here, and it'd be great to have you contribute a fix. IMO this fixes a pretty important bug so it'd be awesome to get this in the next 2.4 release.

**Comment by emergentcypher** [comment]
Hey, sorry, been awol due to moving apartments and such. Been a busy month. 

I've started working on it again and will update the PR soon!

File: framework/src/play/src/main/scala/play/api/mvc/WebSocketApi.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import play.core.Execution.Implicits.internalContext
+import WebSocket._
+
+trait WebSocketApi {
+
+  /**
+   * The WebSocket API stores a reference to an ActorSystem on which it will
+   * create actors to handle websocket connections.
+   */
+  val actorSystem: ActorSystem
```
Comment (richdougherty) [inline framework/src/play/src/main/scala/play/api/mvc/WebSocketApi.scala]: Maybe make this a `def` so we don't limit how the `ActorSystem` is resolved?

**Comment by emergentcypher** [comment]
A bit of an update but would like some guidance, don't know if it'll pass the build right now:
* I presume we're deprecating just the WebSocket object, but not the class, which is used by the builder's methods? Especially if you just want to handle WebSockets with iteratees and not care about actor systems.
* Is WebSocketBuilder ok as a trait, with a Default class, or should it be a case class itself?
* What should be done with the FrameFormatter trait and object? Right now WebSocketBuilder is importing these objects from the (now deprecated) WebSocket object.
* In `WebSocketBuilderModule`, I put the binding for `WebSocketBuilder`. Is it necessary to express its dependency on ActorSystem? I'm less familiar with the runtime DI framework.

Once this is finalized I'll update tests and documentation.

**Comment by gmethvin** [comment]
@emergentcypher I think the CI is broken right now so don't worry about the build failure yet.

We should deprecate the specific methods in the `WebSocket` object rather than the object itself. There's no problem with using the object itself for things where no state is involved. You can leave `FrameFormatter` in there. To deprecate those methods you should use scala's `@deprecated` (there are examples elsewhere in Play).

I think a trait makes sense. It makes it easy to create a custom implementation if needed.

It's probably best to just specify the binding in the `BuiltInModule` instead of creating a module for it. You should not need to specify the dependency on `ActorSystem` in the module.

You also should add some tests using the builder in `WebSocketSpec`.

**Comment by benmccann** [comment]
@emergentcypher just a reminder that you have this PR out. it'd be helpful to rebase it and make the previously mentioned changes if you want to continue with it

**Comment by gmethvin** [comment]
It looks like this PR has been abandoned (the user that created it no longer exists).

**Comment by wsargent** [comment]
Closing.

**Pull request closed without merging**